src/renderer/Dataframe.js

```diff
@@ -134,28 +139,127 @@ export default class Dataframe {
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.ALPHA,
+            width, height, 0, gl.ALPHA, gl.FLOAT,
+            propertiesFloat32Array);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
+        return this.propertyTex[propertyName];
+    }
+
+    free () {
+        if (this.propertyTex) {
+            const gl = this.renderer.gl;
+            this.propertyTex.map(tex => gl.deleteTexture(tex));
+            gl.deleteTexture(this.texColor);
+            gl.deleteTexture(this.texStrokeColor);
+            gl.deleteTexture(this.texWidth);
+            gl.deleteTexture(this.texStrokeWidth);
+            gl.deleteTexture(this.texFilter);
+            gl.deleteBuffer(this.vertexBuffer);
+            gl.deleteBuffer(this.featureIDBuffer);
+        }
+        const freeObserver = this.freeObserver;
+
+        Object.keys(this).map(key => {
+            this[key] = null;
+        });
+
+        this.freed = true;
+
+        if (freeObserver) {
+            freeObserver(this);
+        }
+    }
+
+    _geometryInViewport (featureIndex, renderScale, center, aspect, viz) {
+        const feature = this.getFeature(featureIndex);
+        let strokeWidthScale = 1;
 
         switch (this.type) {
-            case 'point': {
-                const x = this.geom[2 * featureIndex + 0];
-                const y = this.geom[2 * featureIndex + 1];
-                return x > minx && x < maxx && y > miny && y < maxy;
-            }
+            case 'point':
+                return this._isPointInViewport(featureIndex, renderScale, center, aspect);
             case 'line':
-            case 'polygon': {
-                const aabb = this._aabb[featureIndex];
-                return !(minx > aabb.maxx || maxx < aabb.minx || miny > aabb.maxy || maxy < aabb.miny);
-            }
+                strokeWidthScale = this._computeLineWidthScale(feature, viz);
+                return this._isPolygonInViewport(featureIndex, renderScale, strokeWidthScale, center, aspect);
+            case 'polygon':
+                strokeWidthScale = this._computePolygonWidthScale(feature, viz);
+                return this._isPolygonInViewport(featureIndex, renderScale, strokeWidthScale, center, aspect);
             default:
                 return false;
         }
     }
 
-    _getBounds (scale, center, aspect) {
-        this.vertexScale = [(scale / aspect) * this.scale, scale * this.scale];
-        this.vertexOffset = [(scale / aspect) * (center.x - this.center.x), scale * (center.y - this.center.y)];
+    _isPointInViewport (featureIndex, renderScale, center, aspect) {
+        const { minx, maxx, miny, maxy } = this._getBounds(renderScale, center, aspect);
```

Comment:
calling `_getBounds` per point seems like a waste of CPU time in a hot spot

Reply:
It was not calling `_getBounds` per point I think, but I have refactored this method by passing the viewportAABB (bounds) as a parameter.